Cirrious/Cirrious.MvvmCross/ViewModels/MvxAsyncCommand.cs

```diff
@@ -0,0 +1,283 @@
+﻿using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Cirrious.CrossCore;
+using Cirrious.CrossCore.ExtensionMethods;
+using Cirrious.CrossCore.Platform;
+
+namespace Cirrious.MvvmCross.ViewModels
+{
+    public abstract class MvxAsyncCommandBase
+        : MvxCommandBase
+    {
+        private readonly object _syncRoot = new object();
+        private readonly bool _allowConcurrentExecutions;
+        private CancellationTokenSource _cts;
+        private int _concurrentExecutions;
+
+        protected MvxAsyncCommandBase(bool allowConcurrentExecutions = false)
+        {
+            _allowConcurrentExecutions = allowConcurrentExecutions;
+        }
+
+        public event EventHandler<MvxCommandErrorEventArgs> ErrorOccured;
+
+        public bool IsRunning
+        {
+            get { return _concurrentExecutions > 0; }
+        }
+
+        protected CancellationToken CancelToken 
+        {
+            get 
+            {
+                return _cts.Token; 
+            }
+        }
+
+        protected abstract bool DoCanExecute(object parameter);
+        protected abstract Task DoExecuteAsync(object parameter);
+
+        public void Cancel()
+        {
+            lock (_syncRoot)
+            {
+                if (_cts == null)
+                {
+                    Mvx.Trace(MvxTraceLevel.Warning, "MvxAsyncCommand : Attempt to cancel a task that is not running");
+                }
+                else
+                {
+                    _cts.Cancel();
+                }
+            }
+        }
+
+        public bool CanExecute()
+        {
+            return CanExecute(null);
+        }
+
+        public bool CanExecute(object parameter)
+        {
+            if (!_allowConcurrentExecutions && IsRunning)
+                return false;
+            else
+                return DoCanExecute(parameter);
+        }
+
+        public async void Execute(object parameter)
+        {
+            await ExecuteAsync(parameter).ConfigureAwait(false);
+        }
+
+        public async void Execute()
```

Comment:
I got that we need synchronous `Execute` methods, while we need `ICommand` interface to be implemented.
Related to synchronous command, as I know, there are no any async code. This means that only library users hold responsibility on handling of exceptions thrown by their code. Introducing `async void` methods, the library takes this responsibility, because users cannot catch that exceptions anymore.
So, even if we can neither rethrow exceptions asyncronously nor return them through an event, I believe all exceptions (including those thrown in `CanExecute` method) should be catched and at least traced as in current implementation.